### PR TITLE
Dev: add a function helper for #14650: Really throw error when user try to hack server

### PR DESCRIPTION
**Pull request by Shnoulle**

### CODE REVIEW

**Comment by Shnoulle** [comment]
It's just a base idea : adding unlink, file_get_contents  etc … all goes to App() with same restriction.

After need to find all rmdirr call (but a grep and it's done)

**Comment by Shnoulle** [comment]
See https://github.com/LimeSurvey/LimeSurvey/commit/7b14cba6f40a6d1e226a3266de48b2d5d52be9e8#diff-f5392477bcb7ace25c65e8bd7abd39e9R384 for `is_file`

**Comment by Shnoulle** [comment]
When error happen : 

1. always log it as error, then server admin can see something happen
2. If user is a super admin + debug : throw error (debug set or not) by default
3. In some condition : throw error  (sample : user is allowed to edit theme, and theme try to remove a bad file)
4. In some condition : throw error (for rmdir for example)

**Comment by c-schmitz** [comment]
Shouldn't all rrmdir calls use the new function and remove the old one?

**Comment by Shnoulle** [comment]
Yes, it's why it's currently an idea , and need to be improved , update other accessand used in future …

I make it after a security issue about template deletion where any user can remove any dir in server … it's fixed but without alert … Try to hack server => need a log way somewhezre (and in my opinion HTTP error is better).

Related discussion about HTML errors : https://github.com/LimeSurvey/LimeSurvey/commit/9a8a03141d45dc12031aa36e8473fca8172c0b37#comments 
Related mantis about HTML errors : https://bugs.limesurvey.org/view.php?id=16470

**Comment by c-schmitz** [comment]
@Shnoulle Can you go ahead and implement this for all rrmdir calls? 
It probably should also go to /dev branch.

**Comment by Shnoulle** [comment]
>  It probably should also go to /dev branch.

Yes , clearly :)

Maybe in https://github.com/LimeSurvey/LimeSurvey/blob/master/application/core/LSFileHelper.php and not in LS_Appliction ?

My opinion : we must create a file loader helper too :)

**Comment by c-schmitz** [comment]
I don't think the right place to place this function would be in the LSYii_Application class.
@olleharstedt what do you think?

**Comment by Shnoulle** [comment]
I leave it as draft, feel free for discussion.

I surely rewrite all when i have time (some plugin make me work hard currently)

**Review by Shnoulle — COMMENTED** [review]
> I don't think the right place to place this function would be in the LSYii_Application class.
> @olleharstedt what do you think?

> Maybe in https://github.com/LimeSurvey/LimeSurvey/blob/master/application/core/LSFileHelper.php and not in LS_Appliction ?

Unsue , FileHelper name is clear, but don't seems totally related.

**Comment by olleharstedt** [comment]
We already have rmdirr. Do we need two? :thinking:

**Comment by olleharstedt** [comment]
Oh, it's a wrapper for rmdirr, kind of. Let me read again.

**Comment by Shnoulle** [comment]
Yep : it's a wrapper 

When fixing : https://bugs.limesurvey.org/view.php?id=14617 **Arbitrary FIle Download in LimeSurvey** 

I think the fix solution are bad : because it's fix ONLY for this call. Then adding a wrapper to

1. rmdir
2.  readfile
3. all file action

Dev must use, if not : it's an issue. 

And in this wrapper : we can clearly send a Ecxeption : then server manager can have log of bad action.

**Comment by Shnoulle** [comment]
See the «fix»

https://github.com/LimeSurvey/LimeSurvey/commit/1ed10d3c423187712b8f6a8cb2bc9d5cc3b2deb8

File: application/core/LSYii_Application.php [hunk 1]
```diff
@@ -180,6 +180,26 @@ public function loadHelper($helper)
...
+     * Throw Exception
+     * @return boolean
+     */
+    public function rmdirr($dirPath,$baseDir = null)
+    {
+        if(empty($baseDir)) {
+            $baseDir = $this->getConfig('uploaddir');
+        }
+        $dirPath = realpath($dirPath);
+        if(!is_dir($dirPath) || substr($dirPath, 0, strlen($baseDir)) !== $baseDir) {
```
[the remark this answers is not in the document]
Comment (olleharstedt) [reply application/core/LSYii_Application.php:196]: Is it possible to write a unit-test for this function? One test for happy path, one for failure. It makes it easier to understand the behaviour.

[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/core/LSYii_Application.php:196]: Open to discussion : it's a 2 years old pull request … 

But : decision about places before no ?

**Comment by c-schmitz** [comment]
SO, either we finish this up or we close it. It's been around long enough.

**Pull request closed without merging**